README.md
# Code::Stats extension to Visual Studio Code.

This is a Visual Studio Code extension to send updates to [https://codestats.net](https://codestats.net)

## Features

This extension tracks the amount of changes you make to your files and sends out pulses to [https://codestats.net/api-docs#pulse](https://codestats.net/api-docs#pulse) to track your XP.

## Extension Settings

You need to register your machine at [https://codestats.net/my/machines](https://codestats.net/my/machines) to get an API key which you need to set in the settings.

This extension contributes the following settings:

* `codestats.apikey`: API key for your machine

## Known Issues

## Release Notes

### 1.0.6

Changed to return just the langId if not found in the mapped languages names.
Added Elm and Elixir to the mapped names.

### 1.0.5

Changed language names to reflect the ones already used in Code::Stats.

### 1.0.4

Added a manual mapping to natural language language names.

### 1.0.3

Fixed an accumulation of XP bug.
Added a license for the logo.

### 1.0.2

Added a logo for the extension.

### 1.0.1

Added the missing github information to package.json
Added the .vsix installation package as well.

### 1.0.0

Initial release of code-stats-vscode


### For more information

* [Github repository](https://github.com/riussi/code-stats-vscode)

**Enjoy!**

CHANGELOG.md
# Change Log
All notable changes to the "code-stats-vscode" extension will be documented in this file.

## [Unreleased]
### Added

### Changed

## [1.0.6] - 2017-03-24
### Changed
Changed the plugin to default to just langId if the langId is not found in the mapped language names.
Added Elm and Elixir to mapped names.

## [1.0.5] - 2017-03-18
### Changed
Changed language names to reflect the ones already used in Code::Stats.

## [1.0.4] - 2017-03-18
### Added
Added a manual mapping to natural language language names.

## [1.0.3] - 2017-03-18
### Added
Added license for the logo from Nicd.

### Changed
Fixed an accumulation of xp bug

## [1.0.2] - 2017-03-17
### Added
Added a logo for the extension.

## [1.0.1] - 2017-03-17
### Added
Added the missing github information to package.json
Added the .vsix installation package as well.

## [1.0.0] - 2017-03-17
### Added
Initial release of code-stats-vscode


test/extension.test.ts
import * as assert from "assert";

// tslint:disable-next-line:no-unused-variable
import * as vscode from "vscode";
// tslint:disable-next-line:no-unused-variable
import * as codestats from "../src/code-stats";
import { Pulse } from "../src/pulse";

suite("code-stats-vscode extension tests", () => {
    test("Initialized Pulse is empty", () => {
        let pulse: Pulse = new Pulse();
        const language: string = "typescript";

        let initialXP: number = pulse.getXP(language);

        assert.equal(initialXP, 0);
    });

    test("Add XP to Pulse", () => {
        let pulse: Pulse = new Pulse();

        const language1: string = "typescript";
        const language2: string = "javascript";
        const language3: string = "coffeescript";
        const addedXP: number = 1000;

        let xp1: number = pulse.getXP(language1);
        let xp2: number = pulse.getXP(language2);
        let xp3: number = pulse.getXP(language3);
        assert.equal(xp1, 0);
        assert.equal(xp2, 0);
        assert.equal(xp3, 0);

        pulse.addXP(language1, addedXP);
        xp1 = pulse.getXP(language1);
        xp2 = pulse.getXP(language2);
        xp3 = pulse.getXP(language3);

        assert.equal(xp1, addedXP);
        assert.equal(xp2, 0);
        assert.equal(xp3, 0);

        pulse.addXP(language1, addedXP);
        pulse.addXP(language2, addedXP);
        xp1 = pulse.getXP(language1);
        xp2 = pulse.getXP(language2);
        xp3 = pulse.getXP(language3);

        assert.equal(xp1, 2 * addedXP);
        assert.equal(xp2, addedXP);
        assert.equal(xp3, 0);
    });
});

src/xp-counter.ts
// tslint:disable-next-line:max-line-length
import { Disposable, workspace, window, StatusBarItem, TextDocument, StatusBarAlignment, TextDocumentChangeEvent, Range, WorkspaceConfiguration } from "vscode";
import { Pulse } from "./pulse";
import { CodeStatsAPI } from "./code-stats-api";

export class XpCounter {
    private combinedDisposable: Disposable;
    private statusBarItem: StatusBarItem;
    private pulse: Pulse;
    private api: CodeStatsAPI;
    private updateTimeout: any;

    //    private languages: Array<string> = ["typescript", "javascript"];

    // wait 10s after each change in the document before sending an update
    private UPDATE_DELAY = 10000;

    constructor() {
        this.pulse = new Pulse();

        /* // print out supported language names
        let allLanguages = languages.getLanguages().then(
            (result => {
                console.log(JSON.stringify(result));
            })
        );
        */

        let config: WorkspaceConfiguration = workspace.getConfiguration("codestats");
        if (!config) {
            return;
        }

        // tslint:disable-next-line:typedef
        let apiKey = config.get("apikey");
        this.api = new CodeStatsAPI(`${apiKey}`);

        if (!this.statusBarItem) {
            this.statusBarItem = window.createStatusBarItem(StatusBarAlignment.Left);
        }

        let subscriptions: Disposable[] = [];
        workspace.onDidChangeTextDocument(this.onTextDocumentChanged, this, subscriptions);
        this.combinedDisposable = Disposable.from(...subscriptions);
    }

    dispose(): void {
        this.combinedDisposable.dispose();
        this.statusBarItem.dispose();
    }

    private onTextDocumentChanged(event: TextDocumentChangeEvent): void {
        let changeCount: number = 0;
        for (let change of event.contentChanges) {
            changeCount += this.determineChangeCount(change.range);
        }
        this.updateXpCount(event.document, changeCount);
    }

    private determineChangeCount(range: Range): number {
        if (range === null || range === undefined ) {
            return 0;
        }
        if (range.start.line === range.end.line) {
            if (range.start.character === range.end.character) {
                return 1;
            } else {
                return range.end.character - range.start.character;
            }
        }
        // multiline changes, copy/paste count as 1XP
        return 1;
    }

    public updateXpCount(document: TextDocument, changeCount: number): void {
        let show: boolean;
        if (this.isSupportedLanguage(document.languageId)) {
            this.pulse.addXP(document.languageId, changeCount);
            show = true;
        } else {
            show = false;
        }
        this.updateStatusBar(show, `${this.pulse.getXP(document.languageId)}`);

        // each change resets the timeout so we only send updates when there is a 10s delay in updates to the document
        if (this.updateTimeout !== null) {
            clearTimeout(this.updateTimeout);
        }

        this.updateTimeout = setTimeout(() => {
            this.api.sendUpdate(this.pulse);
        }, this.UPDATE_DELAY);
    }

    private updateStatusBar(show: boolean, changeCount: string): void {
        if (!show) {
            this.statusBarItem.hide();
        } else {
            this.statusBarItem.text = `$(pencil) C::S ${changeCount}`;
            this.statusBarItem.show();
        }
    }

    private isSupportedLanguage(language: string): boolean {
        // todo: check supported languages
        // only update xp if one of supported languages
        return true;
    }

}


src/code-stats-api.ts

import { Pulse } from "./pulse";
import { getISOTimestamp, getLanguageName } from "./utils";
import * as axios from "axios";

export class CodeStatsAPI {
    private API_KEY = null;
    private UPDATE_URL = "https://codestats.net/api/my/pulses";
    private axios;

    constructor(apiKey: string) {
        this.API_KEY = apiKey;
        if (this.API_KEY === null || this.API_KEY === undefined) {
            return;
        }

        this.axios = axios.default.create({
            baseURL: this.UPDATE_URL,
            timeout: 1000,
            headers: {
                "X-API-Token": this.API_KEY,
                "Content-Type": "application/json"
            }
        });

    }

    public sendUpdate(pulse: Pulse): void {
        // tslint:disable-next-line:typedef
        const data = new ApiJSON(new Date());

        for (let lang of pulse.xps.keys()) {
            let languageName: string = getLanguageName(lang);
            let xp: number = pulse.getXP(lang);
            data.xps.push(new ApiXP(languageName, xp));
        }

        let json: string = JSON.stringify(data);
        console.log(`JSON: ${json}`);

        this.axios.post(this.UPDATE_URL, json)
            .then( (response) => {
                console.log(response);
            })
            .then ( () => {
                pulse.reset();
            })
            .catch((error) => {
                console.log(error);
            });
    }
}

class ApiJSON {
    constructor(date: Date) {
        this.coded_at = getISOTimestamp(new Date());
        this.xps = new Array<ApiXP>();
    }

    coded_at: string;
    xps: Array<ApiXP>;
}

class ApiXP {
    constructor(language: string, xp: number) {
        this.language = language;
        this.xp = xp;
    }

    language: string;
    xp: number;
}


src/utils.ts
// converted to ts from https://github.com/Nicd/code-stats-atom/blob/master/lib/utils.js
export function getISOTimestamp(date: Date): string {
    const offset: number = -date.getTimezoneOffset();
    const prefix: string = (offset >= 0) ? "+" : "-";

    function pad(num: number): string {
        const norm: number = Math.abs(Math.floor(num));

        return ((norm < 10) ? "0" : "") + norm;
    }

    return date.getFullYear() +
        "-" + pad(date.getMonth() + 1) +
        "-" + pad(date.getDate()) +
        "T" + pad(date.getHours()) +
        ":" + pad(date.getMinutes()) +
        ":" + pad(date.getSeconds()) +
        prefix + pad(offset / 60) +
        pad(offset % 60);
}

export function getLanguageName(langId: string): string {
    // currently supported language ids in vscode as of 2017-03-18
    let languageNames: { [key:string]: string; } = {
        "plaintext": "Plain text",
        "Log": "Log",
        "bat": "Batch",
        "clojure": "Clojure",
        "coffeescript": "CoffeeScript",
        "c": "C",
        "cpp": "C++",
        "csharp": "C#",
        "css": "CSS",
        "diff": "Diff",
        "dockerfile": "Docker",
        "elixir": "Elixir",
        "elm": "Elm",
        "fsharpcss": "F#",
        "git-commit": "Git",
        "git-rebase": "Git",
        "go": "Go",
        "groovy": "Groovy",
        "handlebars": "Handlebars",
        "hlsl": "HLSL",
        "html": "HTML",
        "ini": "Ini",
        "properties": "Properties",
        "java": "Java",
        "javascriptreact": "JavaScript (React)",
        "javascript": "JavaScript",
        "jsx-tags": "JavaScript (JSX)",
        "json": "JSON",
        "less": "LESS",
        "lua": "Lua",
        "makefile": "Makefile",
        "markdown": "Markdown",
        "objective-c": "Objective-C",
        "perl": "Perl",
        "perl6": "Perl 6",
        "php": "PHP",
        "powershell": "PowerShell",
        "jade": "Pug",
        "python": "Python",
        "r": "R",
        "razor": "Razor",
        "ruby": "Ruby",
        "rust": "Rust",
        "scss": "SCSS",
        "shaderlab": "Shaderlab",
        "shellscript": "Shell Script",
        "sql": "SQL",
        "swift": "Swift",
        "typescript": "TypeScript",
        "typescriptreact": "TypeScript (React)",
        "vb": "Visual Basic",
        "xml": "XML",
        "xsl": "XSL",
        "yaml": "YAML"
    };

    let languageName: string = languageNames[langId];

    if (languageName === null || languageName === undefined) {
        return langId;
    }
    return languageName;
}
